### DISCUSSION TITLE
Version controlling the universal model and the sequences

### DISCUSSION THREAD

**Discussion by Midnighter** [jotech/gapseq#81] [Ideas]
I admit that I'm a bit of a control freak when it comes to versioning but I do think that the following will have some benefits for gapseq and its design.

From a provenance perspective, the ideal scenario is that I know the version of every data input as well as the version of the method applied to that data such that I know exactly how and from what a result was generated. At the moment, the universal model is part of the gapseq installation. So it is at least versioned in git. Allowing the universal model to be an input parameter rather than built in would make gapseq more flexible, though.

My main issue is with the `update_sequences.sh` script which downloads those sequences from your FTP server. Usually, I run this during the installation. In general, I have no control over which version of the sequences gets installed nor do I have any information on what changes when I run the update script at a later point.

These are the two data inputs that I currently see. There may be others as well. So my main suggestion is to provide some form of for the sequences (with a change log) and to enable gapseq to read the universal model, as well as the sequences from input parameters.

I look forward to your thoughts and I'm definitely willing to contribute if you consider this.

**Comment by jotech** [comment]
hi @Midnighter, good point! There are some ways to address this but you are definitely right there should be more ways to control it! What already is implemented:
- sequence database version/hash is written to output files (header of *Pathways.tbl, *Reactions.tbl)
- models have gapseq version/hash (model nodes field of xml)
- sequence files keep timestamp of last uniprot modification (example: ``ls -ltr dat/seq/Bacteria/rev/3.5.1.74.fasta`` coresponds to [[1]](https://www.uniprot.org/uniprot/?query=ec%3A3.5.1.74+taxonomy%3Abacteria&sort=score) and [[2]](https://www.uniprot.org/uniprot/Q9KK62?version=*)
- you could always get the latest sequences from uniprot by using ``gapseq find -U``, but download is quite slow

It think for the next sequence update, we will go for version numbers so that users could choose older version, too!

**Reply by Midnighter** [reply]
All those points sound really good.

I think what I'm also looking for is to download a specific version and provide that as input to `gapseq find` (so that the sequences don't have to be located in `gapseq/dat/seq`). Is that already possible?

**Reply by jotech** [reply]
no that's not yet possible
